src/clj_pgp/message.clj

```diff
@@ -241,70 +256,78 @@
-               iterator-seq
-               (map #(when-let [ds (unpack-message % opts)]
-                       (vector % ds)))
-               (remove nil?)
-               first)]
-      (when-not content
-        (throw (IllegalArgumentException.
-                 (str "Cannot decrypt " (pr-str packet) " with " (pr-str opts)
-                      " (no matching encrypted session key)"))))
-      (when (and (.isIntegrityProtected object)
-                 (not (.verify object)))
-        (throw (IllegalStateException.
-                 (str "Encrypted data object " object
-                      " failed integrity verification!"))))
-      (let [mdc (.isIntegrityProtected object)]
-        (mapv #(assoc % :integrity-protected? mdc) content))))
+    (some
+      #(readable % opts)
+      (iterator-seq (.getEncryptedDataObjects packet))))
 
 
   PGPPBEEncryptedData
 
+  (reduce-message
+    [packet {:keys [decryptor] :as opts} rf acc]
+    (let [decryptor-factory (BcPBEDataDecryptorFactory.
+                              (.toCharArray ^String decryptor)
+                              (BcPGPDigestCalculatorProvider.))
+          cipher (-> packet
+                     (.getSymmetricAlgorithm decryptor-factory)
+                     tags/symmetric-key-algorithm-tag)]
+      (reduce-content
+        (.getDataStream packet decryptor-factory)
+        opts
+        #(rf %1 (assoc %2 :cipher cipher :object packet))
+        acc)))
+
   ;; If the decryptor is a string, try to use it to decrypt the passphrase
   ;; protected session key.
-  (unpack-message
-    [packet opts]
-    (let [decryptor (:decryptor opts)]
-      (when (string? decryptor)
-        (let [decryptor-factory (BcPBEDataDecryptorFactory.
-                                  (.toCharArray ^String decryptor)
-                                  (BcPGPDigestCalculatorProvider.))
-              cipher (-> packet
-                         (.getSymmetricAlgorithm decryptor-factory)
-                         tags/symmetric-key-algorithm-tag)]
-          (mapv #(assoc % :cipher cipher)
-                (-> packet
-                    (.getDataStream decryptor-factory)
-                    (expand-content opts)))))))
+  (readable
+    [packet {:keys [decryptor] :as opts}]
+    (when (string? decryptor)
+      packet))
 
 
   PGPPublicKeyEncryptedData
 
+  (reduce-message
+    [packet {:keys [decryptor] :as opts} rf acc]
+    (let [for-key (.getKeyID packet)
+          privkey (pgp/private-key
+                    (if (ifn? decryptor)
+                      (decryptor for-key)
+                      decryptor))
+          decryptor-factory (BcPublicKeyDataDecryptorFactory. privkey)
+          cipher (-> packet
+                     (.getSymmetricAlgorithm decryptor-factory)
+                     tags/symmetric-key-algorithm-tag)]
+      (reduce-content
+        (.getDataStream packet decryptor-factory)
+        opts
+        #(rf %1 (assoc %2 :encrypted-for for-key :cipher cipher :object packet))
```

Comment:
This anonymous function signature has shown up a few times now - seems like it could be extracted to a helper function:
```clojure
(defn with-reduce-attrs
  [rf & kvs]
  (fn reducer
    [acc value]
    (rf acc (apply assoc value kvs))))
```
Then this line could be:
```clojure
(with-reduce-attrs rf
                   :encrypted-for for-key
                   :cipher cipher
                   :object packet)
```